server/graphql/utils/requestType.ts

```diff
@@ -1,27 +1,66 @@
-import { ServerResponseInterface } from '../serverResponse'
+import { Document, Types } from 'mongoose'
+import { softDeleteRequestHelper } from './request'
 import { updateRequestGroupHelper } from './requestGroup'
 
 import { UserInputError } from 'apollo-server-errors'
 
-const updateRequestTypeHelper = (requestType, dataSources): Promise<ServerResponseInterface> => {
+import mongoose from 'mongoose'
+
+const updateRequestTypeHelper = async (requestType, dataSources): Promise<Document> => {
   if(!requestType.id) {
     throw new UserInputError('Missing argument value', { argumentName: 'id' })
   }
-  const requestGroupId = dataSources.requestTypes.getById(requestType.id.toString()).requestGroup
-  return dataSources.requestTypes.update(requestType)
-    .then(res => {
-      updateRequestGroupHelper({"id": requestGroupId}, dataSources)
-      return res
-    })
-    .catch(error => {
-      return error
-    })
+  const session = await mongoose.startSession()
+  try {
+    session.startTransaction()
+    const currentRequestType = dataSources.requestTypes.getById(requestType.id.toString())
+    const oldRequestGroupId = currentRequestType.requestGroup.toString()
+    const res = await dataSources.requestTypes.update(requestType, session)
+    if(requestType.requestGroup) {
+      const newRequestGroupId = requestType.requestGroup.toString()
+      const oldRequestGroup = dataSources.requestGroups.getById(oldRequestGroupId)
+      const newRequestGroup = dataSources.requestGroups.getById(requestType.requestGroup.toString())
+      oldRequestGroup.requestTypes = oldRequestGroup.requestTypes.filter(id => !id.equals(requestType.id))
+      newRequestGroup.requestTypes.push(requestType.id)
+      await updateRequestGroupHelper({"id": oldRequestGroupId, "requestTypes": oldRequestGroup.requestTypes}, dataSources)
+      await updateRequestGroupHelper({"id": newRequestGroupId, "requestTypes": newRequestGroup.requestTypes}, dataSources)
+      requestType.requestGroup = Types.ObjectId(requestType.requestGroup)
+    }
+    else {
+      await updateRequestGroupHelper({"id": oldRequestGroupId}, dataSources)
+    }
+    await session.commitTransaction()
+    return res
+  }
+  catch(error) {
+    console.log(error)
+    await session.abortTransaction()
+    throw error
+  }
+  finally {
+    session.endSession()
+  }
 }
 
-const softDeleteRequestTypeHelper = (id, dataSources): Promise<ServerResponseInterface> => {
-  const requestType = dataSources.requestTypes.getById(id)
-  requestType.requests.map(id => {dataSources.requests.softDelete(id)})
-  return dataSources.requestTypes.softDelete(id)
+const softDeleteRequestTypeHelper = async (id, dataSources): Promise<Document> => {
+  const session = await mongoose.startSession()
+  try {
```

Comment:
Using session.withTransaction automatically aborts the session if an error is thrown. Wherever possible, please remove all try/catch and replace with session.withTransaction

Reply:
https://mongoosejs.com/docs/transactions.html

Reply:
I didn't use withTransaction because we still need to return values from the transaction. As far as I'm aware returning values from withTransaction is quite finicky so I didn't opt for that. Also since we want to propagate errors that are thrown, I thought it would be better this way since we have more control over the process.

Reply:
From the docs, errors should be thrown to the caller of the function if we use withTransaction. 

The docs also give an example of returning values withTransaction. Just wondering what makes it finicky.

The reason I think withTransaction is preferable is that it is a lot more readable and can be called by the parent. E.g: the parent calls the function softDeleteRequestTypeHelper inside of withTransaction. Then softDeleteRequestTypeHelper only needs to perform the necessary queries.

Reply:
The sample given in the docs doesn't actually return the value outside of the withTransaction call. This issue has been raised before https://jira.mongodb.org/browse/NODE-2014 and provided some workarounds involving setting a variable outside of the withTransaction call. That's why I thought it was finicky and imo less readable than manually controlling the transaction.